## Homework 3
This homework asks you to scrape from three different sources, please follow the instructions and do the best you can. With the exception of the first cell that imports the URL request and the beautiful soup library, I have not pre-written any code for you. Do not be afraid--you may look at the tutorial for examples, as well as the Beautiful Soup documentation, and any other Python resource (such a Stack overflow) if need be. 

This homework is intended to be challenging, if you only get 70% of the stuff done than you did very well.

In [2]:
from urllib.request import urlopen
from bs4 import BeautifulSoup

## Shakespeare 
I've posted some relatively simple HTML of the first act of William Shakespeare's The Tempest here: http://floatingmedia.com/columbia/tempest.html please use that URL to download the HTML and put it through Beautiful Soup. (It also may be very helpful for you to go to that page in chrome and inspect elements why you do this.) 
**Please note: there are 2 scenes in Act 1**



In the cells below you should you make two lines--a variable that requests and reads the HTML from the URL (http://floatingmedia.com/columbia/tempest.html), and a second variable that passes the raw HTML into Beautiful Soup. You will use that beautiful soup variable to search the HTML.

In [28]:
my_url = "http://floatingmedia.com/columbia/tempest.html"
raw_html = urlopen(my_url).read()
soup_doc = BeautifulSoup(raw_html, "html.parser")
# print(soup_doc.prettify())

Get the title of the play:

In [4]:
soup_doc.title.text

'The Tempest'

Get the HTML that contains the setting of Act One Scene 1:

In [29]:
acts = soup_doc.find_all('div')
acts[0].text

# alternative: soup_doc.div
# soup_doc.find("div")

'\nAct I, Scene 1\n\nOn a ship at sea\n'

Get the setting of Act One Scene 1 (as a string):

In [33]:
acts[0].find(class_="stagedir").text.strip()

# alternative: soup_doc.div.strong.string

'On a ship at sea'

Get the setting of scene 2

In [30]:
acts[2].find(class_="stagedir").text

# alternative: acts[2].strong.string

'The island. Before PROSPERO’S cell.'

Get the name of the first character in Scene 1

In [8]:
lines = acts[1].find_all('ul')
lines[0].find('strong').text

# alternative: find_all('div').strong.string

'Master. '

Get a list of all of characters in Scene 1 (repeats are fine, you need a loop here)

In [9]:
lines

[<ul><li class="playtext"><strong>Master. </strong>Boatswain!
 </li></ul>,
 <ul><li class="playtext"><strong>Boatswain. </strong>Here, master: what cheer?
 </li></ul>,
 <ul><li class="playtext"><strong>Master. </strong>Good, speak to the mariners: fall to't, yarely,
  <span class="playlinenum">5</span><br/>or we run ourselves aground: bestir, bestir.
 </li></ul>,
 <ul><li class="playtext"><strong>Boatswain. </strong>Heigh, my hearts! cheerly, cheerly, my hearts!
 <br/>yare, yare! Take in the topsail. Tend to the
  <span class="playlinenum">10</span><br/>master's whistle. Blow, till thou burst thy wind,
 <br/>if room enough!
 <br/><i>[Enter ALONSO, SEBASTIAN, ANTONIO, FERDINAND,]</i>
 <br/>GONZALO, and others]
 </li></ul>,
 <ul><li class="playtext"><strong>Alonso. </strong>Good boatswain, have care. Where's the master?
  <span class="playlinenum">15</span><br/>Play the men.
 </li></ul>,
 <ul><li class="playtext"><strong>Boatswain. </strong>I pray now, keep below.
 </li></ul>,
 <ul><li c

In [38]:
a = 0
characterlist = []

for line in lines:
    characterlist.append(line.strong.text)
    a = a + 1

characterlist

['Master. ',
 'Boatswain. ',
 'Master. ',
 'Boatswain. ',
 'Alonso. ',
 'Boatswain. ',
 'Antonio. ',
 'Boatswain. ',
 'Gonzalo. ',
 'Boatswain. ',
 'Gonzalo. ',
 'Boatswain. ',
 'Gonzalo. ',
 'Boatswain. ',
 'Sebastian. ',
 'Boatswain. ',
 'Antonio. ',
 'Gonzalo. ',
 'Boatswain. ',
 'Mariners. ',
 'Boatswain. ',
 'Gonzalo. ',
 'Sebastian. ',
 'Antonio. ',
 'Gonzalo. ',
 'Antonio. ',
 'Sebastian. ',
 'Gonzalo. ']

**Bonus:** write a script that takes all the duplicates out of the list. (This is a python thing, not a beautiful soup thing.)

In [34]:
new_list = set(characterlist)

# new_list = []

# for line in characterlist:
#    if (line in new_list) == False:
#        new_list.append(line)

new_list

{'Alonso. ',
 'Antonio. ',
 'Boatswain. ',
 'Gonzalo. ',
 'Mariners. ',
 'Master. ',
 'Sebastian. '}

Display every stage direction in _scene 2_

In [12]:
stage_direction = acts[3].find_all(class_='stagedir')
stagedir_list = []

for x in stage_direction:
    stagedir_list.append(x.text.strip())
    
stagedir_list


# soup_doc.find_all('div')[3].find_all('p')

['[Enter PROSPERO and MIRANDA]',
 '[Enter ARIEL]',
 '[Exit]',
 '[Enter CALIBAN]',
 '[Draws, and is charmed from moving]',
 '[Exeunt]']

Get the HTML containing Miranda's first speech in Scene 2

In [44]:
a = acts[3].find('ul')

a.text

"Miranda. If by your art, my dearest father, you have\n 85Put the wild waters in this roar, allay them.\nThe sky, it seems, would pour down stinking pitch,\nBut that the sea, mounting to the welkin's cheek,\nDashes the fire out. O, I have suffered\nWith those that I saw suffer: a brave vessel,\n 90Who had, no doubt, some noble creature in her,\nDash'd all to pieces. O, the cry did knock\nAgainst my very heart. Poor souls, they perish'd.\nHad I been any god of power, I would\nHave sunk the sea within the earth or ere\n 95It should the good ship so have swallow'd and\nThe fraughting souls within her.\n"

Now display those same lines without the speaker's name (need to use siblings and a loop to get this)

In [48]:
a.find('li').strong.find_next_siblings(text = True)

# soup_doc.find_all('div')[3].ul.li.strong.next_siblings
# for line in lines:
#    if line.name is None:
#        print(line.string)

['If by your art, my dearest father, you have\n ',
 'Put the wild waters in this roar, allay them.\n',
 'The sky, it seems, would pour down stinking pitch,\n',
 "But that the sea, mounting to the welkin's cheek,\n",
 'Dashes the fire out. O, I have suffered\n',
 'With those that I saw suffer: a brave vessel,\n ',
 'Who had, no doubt, some noble creature in her,\n',
 "Dash'd all to pieces. O, the cry did knock\n",
 "Against my very heart. Poor souls, they perish'd.\n",
 'Had I been any god of power, I would\n',
 'Have sunk the sea within the earth or ere\n ',
 "It should the good ship so have swallow'd and\n",
 'The fraughting souls within her.\n']

Get the HTML containing the speech after Miranda's

In [15]:
acts[3].find('ul').find_next_siblings()[0].text

"Prospero. Be collected:\nNo more amazement: tell your piteous heart\nThere's no harm done.\n 100"

## Box Office Mojo 
I've posted some slightly simplified HTML of the current worldwide box office leaders here: http://floatingmedia.com/columbia/boxoffice.htm please use that URL to download the HTML and put it through Beautiful Soup. (It also may be very helpful for you to go to that page in chrome and inspect elements why you do this.) 

For this part of the assignment you should focus mostly on tr (table rows) amd td (table cells) -- and search them using find_all('tr')[1] in order to get a specific row, and find_all('td')[2] to get a specific cell.

Just like with the Shakespeare write two lines get the raw HTML and then transform the raw HTML into a beautiful soup variable

In [16]:
my_url2 = "http://floatingmedia.com/columbia/boxoffice.htm "
raw_html2 = urlopen(my_url2).read()
soup_doc2 = BeautifulSoup(raw_html2, "html.parser")
print(soup_doc2.prettify())

<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html lang="en">
 <head>
  <title>
   2017 Yearly Box Office Results - Box Office Mojo
  </title>
  <meta content="2017, year, yearly, box, office, result, list, movie, movies, listing, listings, top movies, all time, film" name="keywords"/>
  <meta content="Yearly box office results for 2017." name="description"/>
  <link charset="utf-8" href="/css/mojo.css?1" media="screen" rel="stylesheet" title="no title" type="text/css"/>
  <link charset="utf-8" href="/css/mojo.css?1" media="print" rel="stylesheet" title="no title" type="text/css"/>
 </head>
 <body>
  <h2>
   <a href="/yearly/?view2=worldwide&amp;p=.htm">
    Yearly Box Office
   </a>
  </h2>
  <h1>
   2017 WORLDWIDE GROSSES
  </h1>
  <p>
  </p>
 </body>
</html>
<table border="0" cellpadding="5" cellspacing="1">
 <tr bgcolor="#dcdcdc">
  <td align="center">
   Rank
  </td>
  <td align="center">
   Title
  </td>
  <td align="cente

Get the text "2017 worldwide grosses"

In [17]:
soup_doc2.h1.text

'2017 WORLDWIDE GROSSES'

Get an HTML list of all of the column names from the header of the table: rank, title, etc...

In [18]:
a = soup_doc2.table.find('tr').find_all('td')

[tag.text for tag in a]

['Rank', 'Title', 'Studio', 'Worldwide', 'Domestic', 'Overseas']

Make a variable called "headers" and set it equal to the list of strings of the column names and header. In the next line write: "headers[1]" your output should be "Title"

In [19]:
headers = [tag.text for tag in a]

headers[1]

'Title'

Get the HTML for the row that contains the #1 movie

In [52]:
soup_doc2.table.find_all('tr')[1].find('b').text

'Beauty and the Beast (2017)'

Get the HTML for the row that contains the #10 movie 

In [21]:
soup_doc2.table.find_all('tr')[10].find('b').text

'xXx: The Return of Xander Cage'

Get a list of the top 10 films (You'll need to use the loop and specify a range for the loop.) The list should look like this:

`Beauty and the Beast (2017)
The Fate of the Furious
Guardians of the Galaxy Vol. 2
Logan
Kong: Skull Island
Pirates of the Caribbean: Dead Men Tell No Tales
The Boss Baby
Fifty Shades Darker
Your Name.
xXx: The Return of Xander Cage`

In [59]:
ten_list = []

for a in range(1,11):
        ten_list.append(soup_doc2.table.find_all('tr')[a].find('b').text)

ten_list

# all_rows = soup_doc.table.find_all('tr')

# for row in all_rows[1:11]:
#    print(row.find_all('td')[1].string)

['Beauty and the Beast (2017)',
 'The Fate of the Furious',
 'Guardians of the Galaxy Vol. 2',
 'Logan',
 'Kong: Skull Island',
 'Pirates of the Caribbean: Dead Men Tell No Tales',
 'The Boss Baby',
 'Fifty Shades Darker',
 'Your Name.',
 'xXx: The Return of Xander Cage']

Display as strings the top 10 films by name and worldwide gross. Like this:

`Beauty and the Beast (2017) $502.2
The Fate of the Furious $223.9
Guardians of the Galaxy Vol. 2 $356.7
Logan $226.2
Kong: Skull Island $167.8
Pirates of the Caribbean: Dead Men Tell No Tales $117.6
The Boss Baby $171.0
Fifty Shades Darker $114.4
Your Name. $4.9
xXx: The Return of Xander Cage $44.9`

In [23]:
ten_list2 = []

for a in range(1,11):
    x = soup_doc2.table.find_all('tr')[a].find_all('b')
    ten_list2.append(x[0].text + "  " + x[1].text)

ten_list2

['Beauty and the Beast (2017)  $1,241.3',
 'The Fate of the Furious  $1,229.4',
 'Guardians of the Galaxy Vol. 2  $818.6',
 'Logan  $612.4',
 'Kong: Skull Island  $565.9',
 'Pirates of the Caribbean: Dead Men Tell No Tales  $510.5',
 'The Boss Baby  $483.6',
 'Fifty Shades Darker  $378.8',
 'Your Name.  $354.4',
 'xXx: The Return of Xander Cage  $346.1']

Display the link for the movie `Logan`:

In [24]:
soup_doc2.table.find_all('tr')[4].a['href']

'http://www.boxofficemojo.com/movies/?id=wolverine2017.htm'

Create a list for `Logan` that looks like this:

`['4', 'Logan', 'Fox', '$612.4', '$226.2', '$386.2']`

In [25]:
logan_list = []

y = soup_doc2.table.find_all('tr')[4].find_all('td')

for a in range(0,6):
    logan_list.append(y[a].text)

# alternatvie: without counter
# for cell in logan:
#    logan_list.append(cell.string)
    
logan_list

['4', 'Logan', 'Fox', '$612.4', '$226.2', '$386.2']

Here is some premade code. If you got your headers list and your logan_list right, this will make you a nice dictionary for Logan.

In [26]:
logan_dict = dict(zip(headers, logan_list))
print(logan_dict)

{'Rank': '4', 'Title': 'Logan', 'Studio': 'Fox', 'Worldwide': '$612.4', 'Domestic': '$226.2', 'Overseas': '$386.2'}


## Advanced Challenge
Pitchfork is no go, so try to get out all of the artists and albums (and synopses) from NPRs 2015 list: http://www.npr.org/2015/12/07/458447213/npr-musics-50-favorite-albums-of-2015

Use as many cells as you would like, and don't forget to make a new URL request and a new beautiful soup variable